### ISSUE TITLE
Translations

### ISSUE THREAD

**Issue by paglias** [HabitRPG/habitica#2399]
This is a work item for the ongoing translations.

**Next:**
- [ ] content.coffee 
- [ ] habitrpg-shared messages
- [ ] server

**Done:**
- [x] Find and use an app that can keep strings updated across all languages (Transifex)
- [x] split locales/en/main.json
- [x] allow string interpolation with lodash _.template
- [x] Replace hard-coded strings on **Views** folder
- [x] Add string interpolation to **Views** folder
- [x] Replace **subscriber** hard-coded strings and add string interpolation
- [x] Replace **/public/js/** hard-coded strings and add string interpolation

**Comment by marcelosizer** [comment]
#2364 has shared/profiles/achievements.jade and shared/profiles/stats.jade.

**Comment by marcelosizer** [comment]
@paglias I was looking on how to replace the strings on `.coffee` files and found another project already doing that:

[Locale folder](https://github.com/codecombat/codecombat/tree/master/app/locale)
[Issue on moving strings to the database](https://github.com/codecombat/codecombat/issues/22)
[Comment on issues when english strings are updated](https://github.com/codecombat/codecombat/issues/65#issuecomment-32478427)

Also found this page of [i18next](http://i18next.com/pages/doc_features.html). Have no idea if it's related but I'm linking it anyway =p.

**Comment by paglias** [comment]
@Sinza- thanks, the main problem with the first link's approach is that since habitrpg-shared is used also on the browser we would have to add all the languages to the compiled javascript file making it very heavy or creating an habitrpg-shared version for every single language

**Comment by marcelosizer** [comment]
@paglias @lefnire is the `new-stuff.jade` file created dynamically? How should the strings be replaced? And how is it supposed to be translated?

**Comment by lefnire** [comment]
it's not created dynamically. Hmm, it might be a tough one to translate since we push them out every week or so, it's almost like a blog. should we skip it?

**Comment by marcelosizer** [comment]
I liked the idea of using the wikia to translate it but at the moment there is no wikia for other languages so it just wouldn't be translated.

**Comment by marcelosizer** [comment]
That's the lvlxx under your avatar. Some languages with different alphabets will probably change that.

**Comment by paglias** [comment]
shared/modals/achievements.jade is done

As @luveluen pointed out different languages will have different word orders causing problems with pieces like this:

```
=env.t('rebirthText1')
| {{user.achievements.rebirths}}
=env.t('rebirthText2')
| {{user.achievements.rebirthLevel}}
=env.t('rebirthText3')
```

So the next step, when we will have moved everything to main.json will be to implement string interpolation with lodash `_.template` allowing for variables to be passed to `env.t()`

**Comment by marcelosizer** [comment]
@paglias I've been slowly doing the modals folder and it's almost over. Did you merge it already? I'll have a lot of things to change here.

Also, the items section can be removed from main.json as it's all in the content.coffee.

**Comment by paglias** [comment]
Oh, sorry, I'll revert the commit immediately, let me know when you're done.

**Comment by marcelosizer** [comment]
@paglias I'll probably be done today as I stopped for some days and yesterday I got back to it. I won't change the new_stuff.jade as there was no decision on how to do that and I won't commit drops.jade as I have another PR on that file and I'm afraid to cause inconsistencies.

And if you want to keep that file just let me know.

**Comment by marcelosizer** [comment]
@paglias views/shared/modals is done.
Habitrpg #2522 and Habitrpg-shared [#110](https://github.com/HabitRPG/habitrpg-shared/pull/110).

The file drops.jade is translated but hasn't been commited to avoid version conflict; the files new_stuff.jade and login.jade hasn't been translated as there's no decision on what to do with the first and the latter is not being used anymore.

**Comment by paglias** [comment]
@Sinza- merged everything :)

**Comment by marcelosizer** [comment]
@luveluen on what file?

**Comment by paglias** [comment]
Fantastic work guys! @Sinza- @luveluen 

I'm gonna merge the latest changes by @luveluen now.

Next I'll add string interpolation, then before we start other languages I would like to split `main.json` do you have any advice?

**Comment by marcelosizer** [comment]
\o/ Yeah!

@paglias will `_.template` be used on the jade files? Will we have to update them again?

About the main.json I was thinking on a file for the most common strings like **general.json** with strings like _cancel_, _ok_, and other that appear on several parts. And split the rest of the main.json by folder and have separate files for large ones like we did with terms and privacy. What do you think?

**Comment by paglias** [comment]
@Sinza- it will be totally options, if instead of having a text split into more strings because of some variable in the middle it will be possible to have only one string injecting the variable in the middle with more or less the same syntax. Anyway I'll be doing this.

**Comment by marcelosizer** [comment]
Ok. If you want some help let me know. Can I split the main.json?

**Comment by paglias** [comment]
We only miss drops.jade, I would like to wait for that one @Sinza- then you can split main.json

**Comment by marcelosizer** [comment]
I have already added drops.jade strings to the main.json file so it could be split right now.

**Comment by paglias** [comment]
Perfect! Then if you want, do it!

2014/1/23 Marcelo Sizer notifications@github.com

> I have already added drops.jade strings to the main.json file so it could
> be split right now.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/issues/2399#issuecomment-33141674
> .

## 

Matteo Pagliazzi - paglias.net

**Comment by marcelosizer** [comment]
What about the js files? For instance this is from `inventoryCtrl.js`:

``` javascript
return alert("You must first complete " + $rootScope.Content.quests[item.previous].text + '.');
```

**Comment by paglias** [comment]
I forgot about that... env.t is avalaible as a global object in the browser
and translations are loaded there too so we can translate those as well.
Should we do them before splitting?

**Comment by marcelosizer** [comment]
I can split it first and then create separate .json files for the .js files. That way people can start translating as there's already a lot to translate.

**Comment by paglias** [comment]
I've removed the translationsfor terms & privacy pages because containing legal text we can't have them translated this way.

**Comment by marcelosizer** [comment]
Ok. Let me know when main.json is good to go.

**Comment by paglias** [comment]
I've added the support for data interpolations, used like this:

`"stringNotFound" : "String '<%= string %>' not found.",`

So if we pass a `string` variable to string not found it will render it:

`env.t('stringNotFound', {string: "a string"})`

This is an example (well, now when a string is not found it says the string it was searching for, i thought it could be useful).

Using it is very simple as you can see, we said we could use it instead of having `stringPart1, stringPart2...` and to allow different constructions for different languages, for example:

``` jade
small=env.t('accept1Terms')
  a(href='/static/terms' target='_blank')=env.t('terms')
  =env.t('accept2Terms')
  a(href='/static/privacy' target='_blank')=env.t('privacy')
  | .
```

it's generated by these strings:

``` json
{
  "accept1Terms" : "By clicking the button below, I agree to the ",
  "terms" : "Terms and Conditions",
  "accept2Terms" : " and the ",
  "privacy" : "Privacy Policy"
}
```

How do you think it should be rewritten making it easier to mantain & translate? 

@Sinza- @luveluen

**Comment by marcelosizer** [comment]
The result of that example should be `String a string not found`? If that's correct we would have to pass a variable as "a string" in order to translate it? Otherwise how would you translate the "a string" here `env.t('stringNotFound', {string: "a string"})`?

Like this?

```
"stringNotFound" : "String '<%= string %>' not found.",
"translatedString" : "a string",

var aString = env.t('translatedString')
env.t('stringNotFound', {string: aString)})
```

Can you give an example with the accept terms?

**Comment by marcelosizer** [comment]
Can we do this?

```
small=env.t('acceptTerms')

{
  "acceptTerms" : "By clicking the button below, I agree to the <a href=\"/static/terms\">Terms and Conditions</a> and the <a href=\"/static/privacy\">Privacy Policy</a>.",
}

```

**Comment by paglias** [comment]
@Sinza- 

"a string" is passed as `string`, it's a js object. The string not found is a different example, it uses this code https://github.com/HabitRPG/habitrpg/blob/95f1dc6e600813e44c7d2d94a3a4a8605f2613f6/src/middleware.js#L192.

A more general example could be 

``` json
{
"sayHello": "Hello <%= name %>!"
}
```

and in the view:

```
env.t('sayHello', {name: "Marcelo"})
```

which will produce:

`Hello Marcelo!`

---

About the terms accpetion string, the point in using string interpolation is to avoid forcing a english-style structure of the phrase allowing at the same time to have reusable data.

I mean if someday the url of the terms page changes, or the layout of the page changes we should not need to change every language file but only the url or the tag. I was thinking at something like this:

```
env.('acceptTerms', {termsUrl: "/static/terms", privacyUrl: "/static/privacy"});

and

{
"acceptTerms": "By clicking the button below, I agree to the <a href=\"<%=termsUrl%>\">Terms and Conditions</a> and the <a href=\"<%=privacyUrl%>\">Privacy Policy</a>."
}
```

But I don't like to have html in the translations.

Another example might be:

```
      =env.t('rebirthText1')
      | {{profile.achievements.rebirths}}
      =env.t('rebirthText2')

and

      "rebirthText1" : "Began ",
      "rebirthText2" : " New Adventures",
```

Which I would like to change in:

```
=env.t('rebirthText', {rebirths: "{{profile.achievements.rebirths}}"})

and

      "rebirthText" : "Began <%= rebirths %> New Adventures"
```

**Comment by paglias** [comment]
Mmm... I'll do some of them and then open a pull request so you can see if it's ok or think it can be done better

**Comment by marcelosizer** [comment]
Ok, got it.

**Comment by marcelosizer** [comment]
PRs #2583 and [habitrpg-shared #119](https://github.com/HabitRPG/habitrpg-shared/pull/119) split the main.json file into many others.
- Next I'll do the drops.jade and then update my first drop PR in order to avoid version conflicts if it gets pulled.
- After that I'll review all the files in order to add string interpolation to them.
- Once that's done I'll look into replacing strings in .js files.

When all that's done people can restart translating! =]

**Comment by paglias** [comment]
Don't worry @luveluen , we're advancing very fast toward the moment you'll be able to translate in some language!

**Comment by marcelosizer** [comment]
String interpolation in jade files has been done. Only js files left now.

**Comment by lefnire** [comment]
@paglias @Sinza- I think we have a ton of already-translated stuff to work with now. Do you think we should start getting some of the translations in, even without things like content.coffee (deal with them later, kinda thing)? Or would you prefer to finish up the checklist before getting translators started?

**Comment by marcelosizer** [comment]
@lefnire I have only 1 js file left and it should be up today. I've talked to @paglias and after that I'll upload the folders of the other languages and open them for translations which should take a minute.

Sadly I don't have the knowledge to deal with content.coffee so I'll look into other minor stuff that can help me develop my skills.

**Comment by paglias** [comment]
I'll move on to content.coffee (and maybe server stuff we talked about with @Sinza- ) after finishing Bootstrap 3 upgrade, for the rest I think we can start translations as soon as js files are ready.

**Comment by marcelosizer** [comment]
In a couple of hours =p

**Comment by marcelosizer** [comment]
Only waiting for @paglias to review the PRs and then people will be able to start translating!

**Comment by paglias** [comment]
@luveluen great, I'm going to merge your pr soon!

@Sinza- 

If you have some time this is what I was thinking about: create an issue called Translations in habitrpg-shared and link it in the trello card where the process for translating a string is explained.

Create an issue for every language called "Translations: Language Name" where people who translate that language can subscribe and where we'll post when an update is needed.

When an update to the en locale is published we might have to change all the other locales strings resetting them to the english version in order to not having different strings' content among different languages, probably not the most exciting user experience but as long as we continue to have so frequent deployment of new site versions we can't wait for strings to be updated.

I'll add the arabic language, someone asked for it on Trello.

**Comment by paglias** [comment]
@Sinza- we're already out of sync.... https://github.com/HabitRPG/habitrpg-shared/commit/619d254cc007d5ca58d559513beb2b4f5f688129#diff-cbdbc2cb085f9413c97b6adf9ef73a9dR42 This is a serious problem

**Comment by paglias** [comment]
There are several services for translations out there that integrates with github like https://webtranslateit.com/en/docs/github-integration/ but they are not free (except for small projects) but it costs 24€ / mo .

Do you know of any free service? even without github integration, the important is to support json files

cc @lefnire

**Comment by marcelosizer** [comment]
@paglias I found two options:
- https://www.transifex.com/plans/ - this one is $19 and has no limit of words to open-source projects, also there is the github here: https://github.com/transifex/transifex
- But this one is free for open-source projects and has support for .json files also :+1: : https://crowdin.net 

Are the habitrpg-shared issues required if we start using one of them?

**Comment by paglias** [comment]
crowdin is a no-go: 

> You do not have Commercial products related to the Open Source project you are requesting a license for.

I found this one http://i18next.com/, it's quite similar to our current workflow and has a web interface that we can install on a server

**Comment by marcelosizer** [comment]
i18next has some pluralization, multiline and other stuff that could be nice too.

**Comment by paglias** [comment]
Ahhhh I hate this.

Switching to i18next would require to refactor, well, everything. If we can
find another solution it would be better, I think.

Free services are irrealistic I suppose and we aren't a true open source
project without commercial interests.

@lefnire I created an opensource project on Transifex under the HabitRPG
name, do you want to register it yourself or is it ok created by me?

I'm not sure we can apply as an open source project given the commercial
part of Habit, I'll ask them.

2014-02-07 Marcelo Sizer notifications@github.com:

> i18next has some pluralization, multiline and other stuff that could be
> nice too.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/issues/2399#issuecomment-34487076
> .

## 

Matteo Pagliazzi - paglias.net

**Comment by marcelosizer** [comment]
@paglias switching to i18next would be hard or too much work? If you're willing to guide me I can do it...

**Comment by paglias** [comment]
The syntax shouldn't be a big problem and I suppose the json files can be
kept as they are even thought it should allow for nested properties which
may be useful like {string: {Name:"", text:""} and access it with
string.name / string.content allowing for better organization. They also
have a client side library, if you want to set it up (there's documentation
on their site) feel free to do it, but if you don't know how to do it while
we wait for a response from transifex I'll try to set it up during the
weekend. Sorry for the poor writing, I'm from mobile :)

**Comment by marcelosizer** [comment]
You'll probably do it before I start to read, but I'll take a look.

**Comment by paglias** [comment]
@lefnire (would like to tag Vicky too but don't know her username):

I got an email back from Transifex, they sent me back to their page about open source projects (http://support.transifex.com/customer/portal/articles/1424772-what-is-considered-an-open-source-project-)

Which states that code has to be freely avalaible with an OSI license. I think we meet both requirements given that we use the GPL license and all the code that has to be translated is avalaible open source.

But I want to know if you think it's ok before going forward with this.

I signed up as paglias and created an HabitRPG organization (https://www.transifex.com/organization/habitrpg/). Tell me if that's ok or I can delete it and leave you create it.

---

If we go ahead with this, @Sinza- I believe we might have to trim all the strings (no spaces at the end or the beginning)

**Comment by marcelosizer** [comment]
No problem.

**Comment by veeeeeee** [comment]
@paglias I think we qualify. @lefnire can link you to the current license file.

If they give you any trouble with the legal wording, check back with me.

Thanks for flagging :)

**Comment by paglias** [comment]
@veeeeeee good to know :) The license file is ok, tomorrow I'll start porting translations to transifex and add people as collaborators.

**Comment by lefnire** [comment]
Yeah, we're GPL v3. @veeeeeee think we're legit even with upcoming private code branches? @paglias sorry for being MIA on all this translations stuff, it's so fast and over my head lol. Is there anything you need me to do?

**Comment by paglias** [comment]
@lefnire don't worry, things are going great and a lot of people are contributing. It's also becoming self-sustainable we just need to upload a file when things changes and then everything is automatic.

There's only a problem, which makes everyone having english as their language if when they sign up the right language is not avalaible. Which means that if I sign up now and italian is not avalaible yet I will have `user.preferences.lang === en`.

For that reason I removed the code about that https://github.com/HabitRPG/habitrpg/commit/094a4be0015f0f0deaaf94a0734193eb40a8beae but we need to run a migration to reset all the `user.preferences.lang` to `undefined`. Can you do that?

**Comment by lefnire** [comment]
yeah, you want me to run that now, or later after more translations are out?

**Comment by paglias** [comment]
You can do it now, I've removed the code that saved the user preferred
language until I find how to solve this so it's ok to run it now
Il 12/feb/2014 18:30 "Tyler Renelle" notifications@github.com ha scritto:

**Comment by lefnire** [comment]
√ .  `db.users.update({},{$unset:{'preferences.lang':1}},{multi:true})`

**Comment by paglias** [comment]
@lefnire I'm a stupid... it's `preferences.language` not `preferences.lang`....

**Comment by veeeeeee** [comment]
I don't think we're translating the private code branches. Are we?

**Comment by paglias** [comment]
No we aren't.

@lefnire it's preferences.language not preferences.lang, sorry.
Il 13/feb/2014 07:45 "veeeeeee" notifications@github.com ha scritto:

**Comment by lefnire** [comment]
@paglias √ `db.users.update({},{$unset:{'preferences.lang':1}},{multi:true})`

@veeeeeee we're unlikely to have much content in the private branches. It's mostly backend code.

**Comment by paglias** [comment]
> @paglias √ `db.users.update({},{$unset:{'preferences.lang':1}},{multi:true})`

language not lang....

**Comment by paglias** [comment]
@veeeeeee everything translated is public now

**Comment by lefnire** [comment]
Oh I'm sorry, I ran the right one but commented with the wrong one

**Comment by veghita** [comment]
Thank you very much for the great job you've done with the Spanish translation! If I found any suggestion, where can I send them?

Thanks.

**Comment by paglias** [comment]
@veghita you can go to transifex where the translations are managed and join the spanish team https://www.transifex.com/projects/p/habitrpg/ where you can comment on strings, open discussions

**Comment by veghita** [comment]
Thanks :)

El 15/05/2014 19:23, Matteo Pagliazzi notifications@github.com escribió:
@veghita you can go to transifex where the translations are managed and join the spanish team https://www.transifex.com/projects/p/habitrpg/ where you can comment on strings, open discussions

---
